grovel/grovel.lisp

```diff
@@ -723,21 +815,42 @@ string."
 ;;; written out by PROCESS-WRAPPER-FILE once everything is processed.
 (defvar *lisp-forms*)
 
-(defun generate-c-lib-file (input-file output-defaults)
-  (let ((*lisp-forms* nil)
-        (c-file (make-c-file-name output-defaults "__wrapper")))
-    (with-open-file (out c-file :direction :output :if-exists :supersede)
-      (with-open-file (in input-file :direction :input)
-        (write-string *header* out)
-        (loop for form = (read in nil nil) while form
-              do (process-wrapper-form out form))))
+(defun djb2 (string)
+  (let ((hash 5381)
+        (wrap (- (expt 2 64) 1)))
+    (loop :for c :across string :do
+       (setf hash (mod (+ (ash hash 5) hash (char-code c))
+                       wrap)))
+    hash))
+
+(defun gen-feature-hash (features)
+  (format nil "~a_~a_~x"
+          (or (operating-system) (software-type))
+          (or (architecture) (machine-type))
+          (djb2 (format nil "~{~a~}" features))))
+
+(defun read-wrapper-spec (input-file)
+  (with-cached-reader-conditionals
+    (with-open-file (in input-file :direction :input)
+      (loop :for form = (read in nil nil) :while form :collect form))))
+
+(defun generate-c-lib-file (input-data c-file)
+  (let ((*lisp-forms* nil))
+    (ensure-directory-pathname (pathname-directory-pathname c-file))
+    (with-open-file (out c-file :direction :output
+                         :if-exists :supersede)
+      (write-string *header* out)
+      (loop :for form :in input-data
+         :do (process-wrapper-form out form)))
```

Comment:
As with the groveller version we read with out [feature-expression](http://clhs.lisp.se/Body/24_aba.htm) caching and pass the forms to `generate-c-lib-file` once we know if we already have a result cached or not